source/faq.txt

```diff
@@ -24,13 +24,22 @@ What Is the Difference Between "connectTimeoutMS", "socketTimeoutMS" and "maxTim
      - Description
    * - connectTimeoutMS
      - 30000
-     - ``connectTimeoutMS`` is a `connection-setting <https://mongodb.github.io/node-mongodb-native/3.5/reference/connecting/connection-settings/>`_ option that sets the time in milliseconds to establish a connection to the MongoDB server before timing out.
+     - ``connectTimeoutMS`` is a :doc:`connection option
```

Comment:
we actually just received feedback for this in the [community forums](https://community.mongodb.com/t/connect-timeout-and-execution-timeout-in-nodejs-driver/21290), it seems that this options is easily confused with gating the behavior of `MongoClient#connect`. I think the language should be updated to indicate that it is specific to individual connections in a connection pool. Specifically, this option defines the amount of time the driver's connection pool will wait before determining a connection to a MongoDB server has failed to connect. You might want to additionally call out that this is not directly related to `MongoClient#connect`, and they should instead see `serverSelectionTimeoutMS` for that.

Reply:
I think I'm confused now after reading this part of your explanation:
> You might want to additionally call out that this is not directly related to MongoClient#connect, and they should instead see serverSelectionTimeoutMS for that. 

From various versions of the NodeJS documentation, I thought this setting controlled the maximum interval of time to wait in which no response has been received from the server (for each connection requested in the driver's connection pool, as you clarified) during a call to connect and that `serverSelectionTimeoutMS` related to waiting for replica sets to connect to one another.

If you get the chance, could you invite me to that community forums conversation that you linked? I get an error message that it doesn't exist or is private after logging in. Thanks.

Reply:
Oops, I updated the link.

The problem here is that people think when they set `connectTimeoutMS`, that value will be used as an upper bound on how long it takes for `MongoClient.connect` to complete. That's not the case. `MongoClient.connect` starts an internal server monitor, that monitor creates a connection and `connectTimeoutMS` is used to determine how long we are willing to wait for _that_ connection to connect to a mongod. `MongoClient.connect` is internally implemented as a "server selection", meaning it will attempt to find a suitable server for up to `serverSelectionTimeoutMS`, and otherwise fail. This is not specific to replica sets.

Reply:
Is there any reading you can recommend on how the internal server monitor works (for personal edification)? 

If I understand correctly, during MongoClient.connect:
1) internal server monitor is created - serverSelectionTimeoutMS timer starts
2) monitor requests a connection from the connection pool - connectTimeoutMS starts
3) connection attempts to communicate with the server. If one fails, does it retry with another?
4) connection is established - serverSelectionTimeoutMS and connectTimeoutMS end
Is this correct?

I'm currently adding a new entry for "serverSelectionTimeoutMS" which might resemble the following to help clarify the FAQ answer:
 
>    * - serverSelectionTimeoutMS
>      - 30000
>      - ``serverSelectionTimeoutMS``is a connection option that sets the
>        number of milliseconds to wait for a successful connection with
>        a MongoDB server when calling ``MongoClient#connect``. The default
>        value allows enough time for a typical server primary election to
>        complete. Use this option to limit the time ``MongoClient#connect``
>        waits before failing.

Reply:
Sure, it's documented in the [Server Discovery and Monitoring specification](https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst). 

I don't think we need to document this here, we just need to say that `connectTimeoutMS` is related to establishing individual TCP connections to a server, and `serverSelectionTimeoutMS` is probably what they're looking for if they are trying to make timing out `MongoClient#connect` faster. You won't find specification documentation for a `connect` method, because most drivers don't have one, it's just something weird that the node driver has.